### PR TITLE
Some leveldb changes that were left on a branch

### PR DESCRIPTION
**Pull request by donequis**
Besides some refactor, it uses the leveldb mempool now, though without optimized values.

### CODE REVIEW

File: rskj-core/src/test/java/co/rsk/datasource/DataSourcePoolTest.java [hunk 1]
```diff
@@ -20,9 +20,12 @@
 
 import org.ethereum.datasource.DataSourcePool;
 import org.ethereum.datasource.KeyValueDataSource;
+import org.ethereum.datasource.LevelDbDataSource;
```
Comment (tinchou) [inline rskj-core/src/test/java/co/rsk/datasource/DataSourcePoolTest.java]: please revert file changes

**Comment by donequis** [comment]
> can multiple threads access these methods concurrently, thus requesting lots of memory?

Not really. The maximum given by this change would be half MB per thread. So not that much.

> are we positive this will bring improvements?
> shouldn't the pool size be dynamic, according to the size of the data we're handling?

We are not sure. These changes were suggested when we were having issues with leveldb, but were never merged. Testing the impact would be good as well as testing changes on the size value. Right now it is not a priority though.

File: rskj-core/src/main/java/org/ethereum/datasource/LevelDbDataSource.java [hunk 2]
```diff
@@ -94,11 +111,13 @@ public void init() {
...
             options.compressionType(CompressionType.NONE);
-            options.blockSize(10 * 1024 * 1024);
-            options.writeBufferSize(10 * 1024 * 1024);
-            options.cacheSize(0);
-            options.paranoidChecks(true);
-            options.verifyChecksums(true);
+            options.blockSize(LDB_BLOCK_SIZE);
+            options.writeBufferSize(LDB_WRITE_BUFFER_SIZE);
+            options.maxOpenFiles(LDB_MAX_OPEN_FILES);
+            options.logger(message -> logger.debug(message));
```
Comment (tinchou) [inline rskj-core/src/main/java/org/ethereum/datasource/LevelDbDataSource.java]: Please use method references: `logger::debug`

File: rskj-core/src/main/java/org/ethereum/datasource/LevelDbDataSource.java [hunk 3]
```diff
@@ -236,23 +251,24 @@ public void delete(byte[] key) {
...
                 logger.trace("~> LevelDbDataSource.keys(): " + name);
             }
+            JniDBFactory.pushMemoryPool(LDB_JNI_MEMORY_POOL_SIZE);
 
             try (DBIterator iterator = db.iterator()) {
                 Set<byte[]> result = new HashSet<>();
-                for (iterator.seekToFirst(); iterator.hasNext(); iterator.next()) {
-                    result.add(iterator.peekNext().getKey());
-                }
+                iterator.seekToFirst();
```
Comment (tinchou) [inline rskj-core/src/main/java/org/ethereum/datasource/LevelDbDataSource.java]: Is this needed? We just created the iterator.

File: rskj-core/src/test/java/co/rsk/datasource/DataSourcePoolTest.java [hunk 4]
```diff
@@ -80,4 +80,5 @@ public void openAndCloseTenTimes() {
             DataSourcePool.closeDataSource("test4");
         }
     }
+
```
Comment (tinchou) [inline rskj-core/src/test/java/co/rsk/datasource/DataSourcePoolTest.java]: Revert unnecessary file change

File: rskj-core/src/test/java/org/ethereum/datasource/LevelDbDataSourceTest.java [hunk 5]
```diff
@@ -19,48 +19,103 @@
...
+        underTest.close();
+
+        File thisDbPath = Paths.get(dbsTempFolder.getRoot().getAbsolutePath(), "to-destroy").toFile();
+
+        assertTrue(thisDbPath.exists());
+        underTest.destroyDB(thisDbPath);
+        assertFalse(thisDbPath.exists());
+    }
+
+    // ---------------------------------------------------------------------------------------------------------------
```
Comment (tinchou) [inline rskj-core/src/test/java/org/ethereum/datasource/LevelDbDataSourceTest.java]: Unnecessary separator

**Comment by rskops** [comment]
SonarQube analysis reported 19 issues
* ![MAJOR][MAJOR] 10 major
* ![MINOR][MINOR] 4 minor
* ![INFO][INFO] 5 info

#### Top 10 issues

1. ![MAJOR][MAJOR] [LevelDbDataSource.java#L1](https://github.com/rsksmart/rskj/blob/39f8ea5742083f52b6eeae37ddeb33ecbbd3cf72/rskj-core/src/main/java/org/ethereum/datasource/LevelDbDataSource.java#L1): Unused field: org.ethereum.datasource.LevelDbDataSource.lastTimeUsed [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=findbugs%3AUUF_UNUSED_FIELD)
1. ![MAJOR][MAJOR] [LevelDbDataSource.java#L71](https://github.com/rsksmart/rskj/blob/39f8ea5742083f52b6eeae37ddeb33ecbbd3cf72/rskj-core/src/main/java/org/ethereum/datasource/LevelDbDataSource.java#L71): Remove this unused "lastTimeUsed" private field. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1068)
1. ![MAJOR][MAJOR] [LevelDbDataSource.java#L117](https://github.com/rsksmart/rskj/blob/39f8ea5742083f52b6eeae37ddeb33ecbbd3cf72/rskj-core/src/main/java/org/ethereum/datasource/LevelDbDataSource.java#L117): java/nio/file/Paths.get(Ljava/lang/String;[Ljava/lang/String;)Ljava/nio/file/Path; reads a file whose location might be specified by user input [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=findsecbugs%3APATH_TRAVERSAL_IN)
1. ![MAJOR][MAJOR] [LevelDbDataSource.java#L118](https://github.com/rsksmart/rskj/blob/39f8ea5742083f52b6eeae37ddeb33ecbbd3cf72/rskj-core/src/main/java/org/ethereum/datasource/LevelDbDataSource.java#L118): java/nio/file/Paths.get(Ljava/lang/String;[Ljava/lang/String;)Ljava/nio/file/Path; reads a file whose location might be specified by user input [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=findsecbugs%3APATH_TRAVERSAL_IN)
1. ![MAJOR][MAJOR] [LevelDbDataSource.java#L118](https://github.com/rsksmart/rskj/blob/39f8ea5742083f52b6eeae37ddeb33ecbbd3cf72/rskj-core/src/main/java/org/ethereum/datasource/LevelDbDataSource.java#L118): Method org.ethereum.datasource.LevelDbDataSource.init() appears to call the same method on the same object redundantly [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3APRMC_POSSIBLY_REDUNDANT_METHOD_CALLS)
1. ![MAJOR][MAJOR] [LevelDbDataSource.java#L120](https://github.com/rsksmart/rskj/blob/39f8ea5742083f52b6eeae37ddeb33ecbbd3cf72/rskj-core/src/main/java/org/ethereum/datasource/LevelDbDataSource.java#L120): java/nio/file/Paths.get(Ljava/lang/String;[Ljava/lang/String;)Ljava/nio/file/Path; reads a file whose location might be specified by user input [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=findsecbugs%3APATH_TRAVERSAL_IN)
1. ![MAJOR][MAJOR] [LevelDbDataSource.java#L243](https://github.com/rsksmart/rskj/blob/39f8ea5742083f52b6eeae37ddeb33ecbbd3cf72/rskj-core/src/main/java/org/ethereum/datasource/LevelDbDataSource.java#L243): Method org.ethereum.datasource.LevelDbDataSource.keys() passes a concatenated string to SLF4J's format string [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3ALO_APPENDED_STRING_IN_FORMAT_STRING)
1. ![MAJOR][MAJOR] [LevelDbDataSource.java#L300](https://github.com/rsksmart/rskj/blob/39f8ea5742083f52b6eeae37ddeb33ecbbd3cf72/rskj-core/src/main/java/org/ethereum/datasource/LevelDbDataSource.java#L300): Either log or rethrow this exception. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1166)
1. ![MAJOR][MAJOR] [LevelDbDataSource.java#L303](https://github.com/rsksmart/rskj/blob/39f8ea5742083f52b6eeae37ddeb33ecbbd3cf72/rskj-core/src/main/java/org/ethereum/datasource/LevelDbDataSource.java#L303): Define and throw a dedicated exception instead of using a generic one. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS00112)
1. ![MAJOR][MAJOR] [LevelDbDataSource.java#L324](https://github.com/rsksmart/rskj/blob/39f8ea5742083f52b6eeae37ddeb33ecbbd3cf72/rskj-core/src/main/java/org/ethereum/datasource/LevelDbDataSource.java#L324): Either log or rethrow this exception. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1166)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Pull request closed without merging**